### PR TITLE
Fix for #171 (Shortcuts not working propertly in Android 13)

### PR DESCRIPTION
**Pull request by Ismael034**
Hello,

I have made a fix for issue #171. To fix the issue I have created a custom intent. The "LaunchActivity" activity will be started when the user clicks the shortcut. Then, the shortcut activity will be launched (included as extra data in the intent).

Also, I created a custom style to get a smoother animation when launching the shortcut.

Thanks

### CODE REVIEW

**Comment by drogga** [comment]
Nice. Thank YOU ! (I don't have anything with A13, so I can't test)
@butzist ^

**Comment by drogga** [comment]
Nice. Thank YOU ! (I don't have anything with A13, so I can't test)
@butzist ^, merging seems no longer possible, without reviewing it first for some reason..., I can (blindly) do that, but I don't understand the code, so...

File: ActivityLauncherApp/src/main/java/de/szalkowski/activitylauncher/LauncherActivity.java [hunk 1]
```diff
@@ -0,0 +1,101 @@
...
+     *
+     * @param component component name
+     * @return true, if valid
+     */
+    private static boolean validateComponentName(String component) {
+        Pattern p = Pattern.compile("^[./a-zA-Z0-9]+$");
+        Matcher m = p.matcher(component);
+        return m.matches();
+    }
+}
```
Comment (butzist) [inline ActivityLauncherApp/src/main/java/de/szalkowski/activitylauncher/LauncherActivity.java:101]: The static methods don't have to be moved inside the new activity. I would suggest to keep them in the third-party folder, but I like the idea of moving then into their own class.

File: ActivityLauncherApp/src/main/AndroidManifest.xml [hunk 2]
```diff
@@ -43,10 +43,20 @@
...
             <intent-filter>
                 <action android:name="android.intent.action.VIEW" />
             </intent-filter>
         </activity>
+        <activity
+            android:name=".LauncherActivity"
+            android:exported="true"
+            android:theme="@style/AppTheme.NoDisplay">
+            <intent-filter>
+                <action android:name="activitylauncher.intent.action.LAUNCH_SHORTCUT"/>
```
Comment (butzist) [inline ActivityLauncherApp/src/main/AndroidManifest.xml:57]: Why new activity? Does android 13 forbid crafting shortcuts die other apps, and this we use the root launcher approach for all shortcuts now?

File: ActivityLauncherApp/src/main/java/org/thirdparty/LauncherIconCreator.java [hunk 3]
```diff
@@ -228,21 +155,22 @@ private static void doCreateShortcut(Context context, String appName, Intent int
...
 
     @TargetApi(26)
-    private static void doCreateShortcut(Context context, String appName, Drawable draw, Intent intent) {
+    private static void doCreateShortcut(Context context, String appName, Drawable draw, Intent extraIntent) {
         ShortcutManager shortcutManager = Objects.requireNonNull(context.getSystemService(ShortcutManager.class));
 
         if (shortcutManager.isRequestPinShortcutSupported()) {
             Icon icon = getIconFromDrawable(draw);
-            intent.setAction(Intent.ACTION_CREATE_SHORTCUT);
+            Intent intent = new Intent("activitylauncher.intent.action.LAUNCH_SHORTCUT");
```
Comment (butzist) [inline ActivityLauncherApp/src/main/java/org/thirdparty/LauncherIconCreator.java:163]: Put into constant

File: ActivityLauncherApp/src/main/res/values/themes.xml [hunk 4]
```diff
@@ -1,6 +1,14 @@
-<?xml version="1.0" encoding="utf-8"?>
-<resources>
-    <style name="AppTheme" parent="Theme.AppCompat.DayNight">
-        <item name="dialogCornerRadius">24dp</item>
-    </style>
-</resources>
+<?xml version="1.0" encoding="utf-8"?>
```
Comment (butzist) [inline ActivityLauncherApp/src/main/res/values/themes.xml:1]: This change seems unrelated to the shortcut creation issue. It's it necessary? Maybe put in separate commit/PR?

**Review by butzist — COMMENTED** [review]
Thank you for submitting this nice fix. Please have a look at the comments that I left and let me know if you want to take care of them or if I should handle them.

File: ActivityLauncherApp/src/main/java/de/szalkowski/activitylauncher/LauncherActivity.java (hunk 1 above)
Comment (Ismael034) [reply ActivityLauncherApp/src/main/java/de/szalkowski/activitylauncher/LauncherActivity.java:101]: So, I did that because it didn't make much sense to me to have the launch activity methods inside a shortcut-related class. Can leave it as it was before but I was a little confused. I can just create a new class in the third party folder with all the launch related methods if that's better

File: ActivityLauncherApp/src/main/AndroidManifest.xml (hunk 2 above)
Comment (Ismael034) [reply ActivityLauncherApp/src/main/AndroidManifest.xml:57]: So, I don't really know why some activities do work and some don't. At first, I thought it was because the activity had $ in the name, but that is not the case. So basically no idea. Launch works fine and when using shortcuts it says the application doesn't exist. I have checked the intent we are setting with the debugger and nothing seems off.

File: ActivityLauncherApp/src/main/res/values/themes.xml (hunk 4 above)
Comment (Ismael034) [reply ActivityLauncherApp/src/main/res/values/themes.xml:1]: Yes. When launching the activity, I could have extended FragmentActivity like it was for the root launcher and leave like that. The issue is that, when clicking on the shortcut, you can see for a brief moment, the actual Activity Launcher app. So instead we can extended AppCompatActivity with a custom theme so it will start smoothly

File: ActivityLauncherApp/src/main/java/de/szalkowski/activitylauncher/LauncherActivity.java (hunk 1 above)
Comment (Ismael034) [reply ActivityLauncherApp/src/main/java/de/szalkowski/activitylauncher/LauncherActivity.java:101]: Maybe create a "Launcher" class and then rename the LauncherActivity class to "ShortcutLauncher"?

Comment (butzist) [reply ActivityLauncherApp/src/main/java/de/szalkowski/activitylauncher/LauncherActivity.java:101]: Yes, thirdparty/Launcher.java or so. The name for the activity is good.

**Review by butzist — APPROVED** [review]

**Comment by butzist** [comment]
Looks good to me. Thanks a lot!

File: ActivityLauncherApp/src/main/java/de/szalkowski/activitylauncher/LauncherActivity.java (hunk 1 above)
Comment (ABOADAMm) [reply ActivityLauncherApp/src/main/java/de/szalkowski/activitylauncher/LauncherActivity.java]: e5aed5879d09521dc284ae2866ba8fbd2ff3cd7a
